src/conversation_engine/engine.py

```diff
@@ -308,102 +397,170 @@ def _handle_additional_info(self, text):
             rospy.loginfo("Action succeeded")
-            self._robot_to_user_pub.publish(" ".join(task_outcome.messages))
+            self._on_task_successful(" ".join(task_outcome.messages))
 
             self._start_new_conversation()
-
         elif task_outcome.result == TaskOutcome.RESULT_MISSING_INFORMATION:
             rospy.loginfo("Action needs more info from user")
 
             sentence = "".join(task_outcome.messages)
-            self._state.wait_for_user(target=self._get_grammar_target(task_outcome.missing_field),
+            target = self._get_grammar_target(task_outcome.missing_field)
+            self._state.wait_for_user(target=target,
                                       missing_field=task_outcome.missing_field)
-            example = self._parser.get_random_sentence(self._state.target)
-            sentence += " For example: '{}'".format(example)
-            self._robot_to_user_pub.publish(sentence)
+            self._on_request_missing_information(sentence, self._grammar, target)
 
         elif task_outcome.result == TaskOutcome.RESULT_TASK_EXECUTION_FAILED:
             rospy.loginfo("Action execution failed")
-            self._robot_to_user_pub.publish("".join(task_outcome.messages))
+
+            self._on_task_outcome_failed("".join(task_outcome.messages))
 
             self._start_new_conversation()
 
         elif task_outcome.result == TaskOutcome.RESULT_UNKNOWN:
             rospy.loginfo("Action result: unknown")
-            self._robot_to_user_pub.publish("".join(task_outcome.messages))
+            self._on_task_outcome_unknown("".join(task_outcome.messages))
 
             self._start_new_conversation()
 
         else:
             rospy.loginfo("Action result: other")
-            self._robot_to_user_pub.publish("".join(task_outcome.messages))
+            self._say_to_user("".join(task_outcome.messages))
 
             self._start_new_conversation()
 
     def _feedback_cb(self, feedback):
         self._latest_feedback = feedback
 
         rospy.loginfo(feedback.current_subtask)
-        self._robot_to_user_pub.publish(describe_current_subtask(feedback.current_subtask))
 
-    def _get_grammar_target(self, missing_field_path):
+    @staticmethod
+    def _get_grammar_target(missing_field_path):
+        """Determine which grammar target to use when a particular field of info is missing."""
         deepest_field_name = missing_field_path.split('.')[-1]
 
+        grammar_target = "T"
+
         if 'location' in deepest_field_name:
-            return 'ROOM_OR_LOCATION'
+            grammar_target = 'ROOM_OR_LOCATION'
+        elif 'entity' in deepest_field_name:
+            grammar_target = 'ROOM_OR_LOCATION'
+        elif 'target' in deepest_field_name:
+            grammar_target = 'NAMED_PERSON'
+        elif 'sentence' in deepest_field_name:
+            grammar_target = 'SAY_SENTENCE'
+        elif 'object' in deepest_field_name:
+            grammar_target = 'NAMED_OBJECT'
```

Comment:
This switch statement feels dirty to me. Let's discuss design soon.

Reply:
This should indeed be a parameter

Reply:
There already is an issue for this: https://github.com/tue-robotics/conversation_engine/issues/16